File: library/src/main/java/com/jwoolston/android/uvc/interfaces/streaming/MJPEGVideoFormat.java
package com.jwoolston.android.uvc.interfaces.streaming;

import android.support.annotation.NonNull;
import com.jwoolston.android.uvc.util.Hexdump;

import timber.log.Timber;

/**
 * @author Jared Woolston (Jared.Woolston@gmail.com)
 */
public class MJPEGVideoFormat extends VideoFormat<MJPEGVideoFrame> {

    private static final int LENGTH = 11;

    private static final int bFormatIndex = 3;
    private static final int bNumFrameDescriptors = 4;
    private static final int bmFlags = 5;
    private static final int bDefaultFrameIndex = 6;
    private static final int bAspectRatioX = 7;
    private static final int bAspectRatioY = 8;
    private static final int bmInterlaceFlags = 9;
    private static final int bCopyProtect = 10;

    private final boolean fixedSampleSize;

    public MJPEGVideoFormat(byte[] descriptor) throws IllegalArgumentException {
        super(descriptor);
        if (descriptor.length < LENGTH) {
            throw new IllegalArgumentException("The provided descriptor is not long enough for an MJPEG Video Format.");
        }
        setFormatIndex((0xFF & descriptor[bFormatIndex]));
        setNumberFrames((0xFF & descriptor[bNumFrameDescriptors]));
        fixedSampleSize = descriptor[bmFlags] != 0;
        setDefaultFrameIndex((0xFF & descriptor[bDefaultFrameIndex]));
        setAspectRatioX((0xFF & descriptor[bAspectRatioX]));
        setAspectRatioY((0xFF & descriptor[bAspectRatioY]));
        setInterlaceFlags(descriptor[bmInterlaceFlags]);
        setIsCopyProtect(descriptor[bCopyProtect] != 0);
    }

    public void addMJPEGVideoFrame(@NonNull MJPEGVideoFrame frame) {
        Timber.d("Adding video frame: %s", frame);
        getVideoFrames().add(frame);
    }

    public boolean getFixedSampleSize() {
        return fixedSampleSize;
    }

    @Override
    public String toString() {
        return "MJPEGVideoFormat{" +
               "formatIndex=" + getFormatIndex() +
               ", numberFrames=" + getNumberFrames() +
               ", fixedSampleSize=" + fixedSampleSize +
               ", defaultFrameIndex=" + getDefaultFrameIndex() +
               ", AspectRatio=" + getAspectRatioX() + ":" + getAspectRatioY() +
               ", interlaceFlags=0x" + Hexdump.toHexString(getInterlaceFlags()) +
               ", copyProtect=" + getIsCopyProtect() +
               '}';
    }
}

File: library/src/main/java/com/jwoolston/android/uvc/interfaces/streaming/MJPEGVideoFormat.kt
package com.jwoolston.android.uvc.interfaces.streaming

import com.jwoolston.android.uvc.streaming.FrameSampleFactory
import com.jwoolston.android.uvc.streaming.VideoSampleFactory
import com.jwoolston.android.uvc.util.Hexdump

import timber.log.Timber

/**
 * @author Jared Woolston (Jared.Woolston@gmail.com)
 */
class MJPEGVideoFormat @Throws(IllegalArgumentException::class)
constructor(descriptor: ByteArray) : VideoFormat<MJPEGVideoFrame>(descriptor) {

    val fixedSampleSize: Boolean

    override val displayName: String
        get() = "MJPEG"

    override fun getSampleFactory(maxPayload: Int, maxFrameSize: Int): VideoSampleFactory {
        return FrameSampleFactory(maxPayload, maxFrameSize)
    }

    init {
        if (descriptor.size < LENGTH) {
            throw IllegalArgumentException("The provided descriptor is not long enough for an MJPEG Video Format.")
        }
        formatIndex = 0xFF and descriptor[bFormatIndex].toInt()
        numberFrames = 0xFF and descriptor[bNumFrameDescriptors].toInt()
        fixedSampleSize = descriptor[bmFlags].toInt() != 0
        defaultFrameIndex = 0xFF and descriptor[bDefaultFrameIndex].toInt()
        aspectRatioX = 0xFF and descriptor[bAspectRatioX].toInt()
        aspectRatioY = 0xFF and descriptor[bAspectRatioY].toInt()
        interlaceFlags = (0xFF and descriptor[bmInterlaceFlags].toInt()).toByte()
        isCopyProtect = (descriptor[bCopyProtect].toInt() != 0)
    }

    fun addMJPEGVideoFrame(frame: MJPEGVideoFrame) {
        Timber.d("Adding video frame: %s", frame)
        videoFrames.add(frame)
    }

    override fun toString(): String {
        return "MJPEGVideoFormat{" +
                "formatIndex=" + formatIndex +
                ", numberFrames=" + numberFrames +
                ", fixedSampleSize=" + fixedSampleSize +
                ", defaultFrameIndex=" + defaultFrameIndex +
                ", AspectRatio=" + aspectRatioX + ":" + aspectRatioY +
                ", interlaceFlags=0x" + Hexdump.toHexString(interlaceFlags) +
                ", copyProtect=" + isCopyProtect +
                '}'.toString()
    }

    companion object {

        private val LENGTH = 11

        private val bFormatIndex = 3
        private val bNumFrameDescriptors = 4
        private val bmFlags = 5
        private val bDefaultFrameIndex = 6
        private val bAspectRatioX = 7
        private val bAspectRatioY = 8
        private val bmInterlaceFlags = 9
        private val bCopyProtect = 10
    }
}
